### ISSUE TITLE
Unpacking in arglist vs inside function

### ISSUE THREAD

**Issue by jariji** [JuliaLang/julia#51107]
Unpacking inside the function assigns into the parent scope; unpacking in the arglist binds a new variable. Should I expect that from something in https://docs.julialang.org/en/v1/manual/variables-and-scoping/ or is that arglist-unpacking a special behavior? 

```jl
let x = 1
    function f(θ)
        (;x) = θ
    end

    θ₀ = (;x=2)
    @show x # 1
    f(θ₀)
    @show x # 2
end

let x = 1
    function f((;x))        
    end

    θ₀ = (;x=2)
    @show x # 1
    f(θ₀)
    @show x # 1
end
```

That's on 1.9.1.

@CameronBieganek says both give `1` then `2` on 1.8.2, implying the behavior changed.

I'm not sure what's supposed to happen here.

[Discourse](https://discourse.julialang.org/t/unpacking-in-arglist-vs-inside-function/103264).

**Comment by CameronBieganek** [comment]
Seems like a bug to me. I think the 1.8 behavior is correct.

**Comment by adienes** [comment]
also only occurs when `f` is defined inside the `let` block

I would imagine this should probably get the `correctness bug` label

**Comment by Keno** [comment]
It's not clear to me which behavior should be correct here. Arguably the second case is just structured arguments, which should introduce a fresh binding. After all, you would expect `function f(x) end` to affect `x` in the parent scope.

**Comment by adienes** [comment]
I think allowing function paramaters to be defined via destructuring is already kind of odd. but it feels most consistent to me if
```
function foo((;x))
    ...
end
```

is equivalent to
```
function foo(t)
    x = t.x
    ...
end
```

so then in the original examples, both should show `x=1` then `x=2` when `f` captures/shadows `x`, and `x=1`, `x=1` when it does not

**Comment by JeffBezanson** [comment]
I do think a case can be made for both behaviors. But it is extremely surprising for a name listed in a formal argument list *not* to introduce a new variable (even if not an argument, it is argument-y). In general, there is a preference for not mutating variables in outer scopes. I would rather require an explicit `=` for that to happen. So frankly I prefer the new behavior.

**Comment by CameronBieganek** [comment]
In Julia 1.8, argument destructuring is inconsistent for tuples and named tuples:

```julia
julia> VERSION
v"1.8.2"

julia> let
           a = 1
           function f((a,)) end
           f((2,))
           a
       end
1

julia> let
           a = 1
           function f((; a)) end
           f((; a=2))
           a
       end
2
```

But in Julia 1.9 they are consistent:

```julia
julia> VERSION
v"1.9.3"

julia> let
           a = 1
           function f((a,)) end
           f((2,))
           a
       end
1

julia> let
           a = 1
           function f((; a)) end
           f((; a=2))
           a
       end

1
```

So, this change is probably a good thing. The manual might need to be tweaked, though. Right now it's a little ambiguous, in particular this part:

> If a function argument name is written as a tuple (e.g. `(x, y)`) instead of just a symbol, then an assignment `(x, y) = argument` will be inserted for you.

**Comment by CameronBieganek** [comment]
If I'm not mistaken, one can imagine that the implicit assignment of values to formal arguments when a function is called is protected by the `local` keyword. So, maybe the manual should be updated to say the same about argument destructuring. In other words, this,

```julia
function f((; a, b))
    # body
end
```

is equivalent to this:

```julia
function f(x)
    local (; a, b) = x
    # body
end
```

**Comment by JeffBezanson** [comment]
Yes absolutely. The more I think about it, these are "pseudo-arguments" and should definitely introduce new variables.
